sdks/java/io/cassandra/src/main/java/org/apache/beam/sdk/io/cassandra/CassandraIO.java

```diff
@@ -0,0 +1,502 @@
+     * {@link #withCassandraService(CassandraService)} or creates and returns an implementation
+     * of a concrete Cassandra service dealing with a Cassandra instance.
+     */
+    @VisibleForTesting
+    CassandraService<T> getCassandraService(PipelineOptions pipelineOptions) {
+      if (cassandraService() != null) {
+        return cassandraService();
+      }
+      return new CassandraServiceImpl<>(connectionConfiguration(), table());
+    }
+
+  }
+
+  @VisibleForTesting
+  static class CassandraSource<T> extends BoundedSource<T> {
+
+    private final Read<T> spec;
+    private final String splitQuery;
+    private final SerializableFunction<PipelineOptions, CassandraService<T>> serviceFactory;
+
+    CassandraSource(Read<T> spec,
+                    String splitQuery,
+                    SerializableFunction<PipelineOptions, CassandraService<T>>
+                               serviceFactory) {
+      this.spec = spec;
+      this.splitQuery = splitQuery;
+      this.serviceFactory = serviceFactory;
+    }
+
+    @Override
+    public Coder<T> getDefaultOutputCoder() {
+      return spec.coder();
+    }
+
+    @Override
+    public void validate() {
+      spec.validate(null);
+    }
+
+    @Override
+    public BoundedReader<T> createReader(PipelineOptions pipelineOptions) {
+      return new CassandraReader(this, serviceFactory.apply(pipelineOptions));
+    }
+
+    @Override
+    public long getEstimatedSizeBytes(PipelineOptions pipelineOptions) throws Exception {
+      return serviceFactory.apply(pipelineOptions).getEstimatedSizeBytes();
+    }
+
+    @Override
+    public List<BoundedSource<T>> splitIntoBundles(long desiredBundleSizeBytes,
+                                                   PipelineOptions pipelineOptions) {
+      return serviceFactory.apply(pipelineOptions)
+          .splitIntoBundles(spec, desiredBundleSizeBytes, serviceFactory);
+    }
+  }
+
+  private static class CassandraReader<T> extends BoundedSource.BoundedReader<T> {
+
+    private final CassandraSource<T> source;
+    private final CassandraService<T> service;
+
+    private CassandraService.Reader<T> reader;
+
+    private CassandraReader(CassandraSource<T> source, CassandraService<T> service) {
+      this.source = source;
+      this.service = service;
+    }
+
+    @Override
+    public boolean start() throws IOException {
+      reader = service.createReader(source.splitQuery, source.spec.entity());
+      return reader.start();
+    }
+
+    @Override
+    public boolean advance() throws IOException {
+      return reader.advance();
+    }
+
+    @Override
+    public void close() {
+      reader.close();
+      reader = null;
+    }
+
+    @Override
+    public BoundedSource getCurrentSource() {
+      return source;
+    }
+
+    @Override
+    public T getCurrent() {
+      return reader.getCurrent();
+    }
+
+    @Override
+    public Instant getCurrentTimestamp() {
```

Comment:
I would expect that people would be storing events with real even-time timestamps in Cassandra. It seems unfortunate to not provide a way to extract them.